### ISSUE TITLE
Provide access to all symbols

### ISSUE THREAD

**Issue by gagern** [Khan/KaTeX#286]
All symbols which are included in the fonts should be accessible by some TeX command. I wrote [some code](https://gist.github.com/gagern/bef7cae7d19cef32b575) to compare the list of symbols (from `symbols.js`) against the fonts (as parsed by [TTX](https://github.com/behdad/fonttools/), resulting in [these files](https://github.com/gagern/KaTeX/commit/86030f7ca43ea3966c5027dafdf44d2ad5c86f03)). Here are my findings:

* [ ] pund sterling sign *£* (in Main_Italic font only)
* [ ] degree sign ° (see https://github.com/gagern/KaTeX/commit/07131ecac9ef88f6be9a6282e4ecb2db77d16243)
* [ ] various combining or modifier diacritics: acute, grave, tilde, dot, diaresis, macron, breve, ring, solidus, caron
* [ ] circumflex ^ and tilde ~ stand-alone
* [ ] em-dash — and en-dash – (will require changes to the lexer, I guess)
* [ ] typographic quotation marks “” and ‘’
* [ ] some misassociated typographic quotes at `\u07e2` and `\u07e3` in the Typewriter font
* [ ] division slash ∕? Not sure we need that, would be interesting to know how that came to be in the font, i.e. whether TeX is using this
* [ ] Maltese cross ✠ in AMS

I ignored some findings found by the script which sounded like something already covered by KaTeX, probably due to some other font using the semantically same symbol with a different code point. I also aggregated similar stuff quite freely, and in the process might have missed some important distinctions. I also ignored stuff which I know is already in progress, the dotless i and j in particular. So the above list is not guaranteed to be complete. Feel free to have a look at the [raw output](https://gist.github.com/gagern/bef7cae7d19cef32b575#file-result-txt) yourself and see whether you can spot anything I missed.

**Comment by kevinb7** [comment]
Nice detective work!

I'm surprised we don't support ```\degree``` yet.  I think the low hanging fruit here is ```\pounds```, ```\degree``` and ```\maltese```.  Could you create a pull request for these?

In terms of the combining marks we support the math mode commands but not the text mode, see http://tex.stackexchange.com/tags/accents/info.  We could probably have a separate issue for accent support in text mode.  From the raw output it looks like the combining marks are being used by KaTeX_Main, but not the other fonts.  I'll update the test cases in https://github.com/Khan/KaTeX/pull/132 to include accents.

Not sure what's up with the Typewriter font.  Not sure how important typographic quotes are.

I didn't realize that en-dash was ```--``` and em-dash was ```---```.  I guess this would be a relatively easy change in the lexer, if we see one ```-``` peek ahead to see if there's another.

**Comment by gagern** [comment]
`\pounds` would benefit from the [`mainit` class](https://github.com/kevinb7/KaTeX/blob/mathrm_mathbb/static/katex.less#L89-92) you declare in #132, so I'd wait on that. Again.

**Comment by kevinb7** [comment]
Sounds good.  I'm spend some more time on https://github.com/Khan/KaTeX/pull/132 this weekend and try to finish it up.

**Comment by kevinb7** [comment]
We're also missing DOUBLE VERTICAL LINE ```\u2016``` which is used for ```\Vert```, ```\lVert```, and ```\rVert```.

**Comment by kevinb7** [comment]
Actually we already support ```\Vert``` which uses ```\u2225``` PARALLEL TO.  We should probably use ```\u2016``` for ```\Vert``` et. al. (including ```\|```) and ```\u2225``` for ```\parallel```.

**Comment by spicyj** [comment]
Normally one writes `^\circ` for a degrees sign, which is presumably why that was missed.

**Comment by gagern** [comment]
> Normally one writes ^\circ for a degrees sign

I try to disabuse people of writing `^\circ` since it results in a different – and in my opinion inferior – rendering. The fact that the font *has* a degree symbol seems like indication that someone agrees. Of course, I'm in the privileged situation of using a German keyboard which has the ° sign readily accessible.

**Comment by kevinb7** [comment]
I think ```^\circ``` is a bit big compared to the degree symbol.  @spicyj Is the a reason for using ```^\circ``` instead of the actual degree symbol?

**Comment by gagern** [comment]
I guess it's a matter of laziness: with LaTeX, getting a proper degree symbol takes some extra packages. You get `\textdegree` fairly easily, but a `\degree` which works in math mode requires `gensymb` or `siunitx`, as far as I know. I assume few people go through all that trouble if `^\circ` works well enough for them. And once the got into that habit, they don't change it if they write for MathJax or KaTeX.

**Comment by kevinb7** [comment]
I had a look at ```gensymb``` and defines ```\degree``` in the following way:
```
\renewcommand{\degree}{\ensuremath{^\circ}}
```
```siunitx``` on the other hand defines it as ```\degree{\SIUnitSymbolDegree}``` and ```\SIUnitSymbolDegree``` as character 176 which is the actual ```\degree``` symbol.

**Comment by gagern** [comment]
@kevinb7 Have a second look, and you will find that `gensymb.sty` actually contains *two* possible definitions for `\degree`. If the `textcomp` package is loaded, it will use the companion font using

```tex
\DeclareMathSymbol{\tcdegree}{\mathord}{gns@font}{176}      % {'260}
\DeclareRobustCommand{\degree}{%
  \ifmmode\tcdegree\else\textdegree\fi}
```

I wasn't aware of this extra requirement; I have `\usepackage{textcomp}` as part of by default preamble. Together with `\usepackage{gensymb}` and `\DeclareUnicodeCharacter{00B0}{\degree}` thanks to [this TeX.SE thread](http://tex.stackexchange.com/q/84469/16923).

**Comment by gagern** [comment]
I had another look at this today. Handling ligatures like `---` or ``` `` ``` in the lexer is probably not fully compatible with TeX: `\def\dash{-} 1 \dash-\dash{} 2` will print an em-dash between the digits in my pdflatex, so the ligature-forming has to happen *after* the lexer. It looks as though TeX did this ligature-forming in its version of the parser. `1 -{-}- 2` will generate three hyphens, so the grouping of characters does affect the outcome.

So one question here is this: do we want to be fully compatible with LaTeX? Or can we reasonably expect our users to enter an emdash as a continuous sequence of three `-`, and thus handle the replacements in question in the lexer?

At first glance, doing things in the lexer would make things easier, since we could just enter the symbols as they are appear in source into the symbol definition. But at second glance, I realized that we have to be careful about modes: a `---` in math mode is a set of three minuses, not an em dash. Likewise, a `''` in math mode is a pair of primes, not a right double quotation mark. So if we do this in the lexer, then we might have to make the lexer (again) more mode-aware than I would like, particularly in the light of #493 which in some places relies on whitespace as the only difference between text and math mode, at least as far as the lexer is concerned.

Where else could we handle ligatures? We could try to include this in the parser, as a post-processing step after a text-mode `parseGroup` or `parseOptionalGroup`. Implementing both optional and required groups using a single function, as done for the special groups, would simplify things here.

**Comment by gagern** [comment]
I had another look at this today. Handling ligatures like `---` or ``` `` ``` in the lexer is probably not fully compatible with TeX: `\def\dash{-} 1 \dash-\dash{} 2` will print an em-dash between the digits in my pdflatex, so the ligature-forming has to happen *after* the lexer. It looks as though TeX did this ligature-forming in its version of the parser. `1 -{-}- 2` will generate three hyphens, so the grouping of characters does affect the outcome.

So I moved the ligature handling into the parser. I'm not opening a pull request yet, since my modifications build on #493, but [here is a preview](https://github.com/gagern/KaTeX/compare/3cf617f3350e8b9aac5d0fd9ac933d7c8cd6aea1...symbols1) of the diff I'd be requesting.

**Comment by gagern** [comment]
Now that #493 has been merged (thanks, Kevin!) I've opened #502 for the mentioned additions. Doesn't take care of *all* the symbols contained in my initial report, but quite a number of them.

**Comment by kevinbarabash** [comment]
"various combining or modifier diacritics" were added in #802 for text mode and we already had `\dot`, `\breve`, `\grave`, etc. for math mode.

**Comment by varunagrawal** [comment]
Hi, thanks for the great work on Katex! It is becoming my go to tool for scientific writing for the web.

I wanted to ask if there is any update on the stand-alone tilde symbole `~`, since a lot of boolean math and probability equations use it. I don't know how hard it would be to include it though.

**Comment by edemaine** [comment]
@varunagrawal Do you mean `\sim`? This has been available for a long time. You can also use `\tilde{}` (probably what was originally meant by this issue) though the spacing seems a bit off.

**Comment by varunagrawal** [comment]
`\sim` would be the right thing! Thank you @edemaine.
